packages/backend/src/nest/registration/registration.functions.ts

```diff
@@ -37,252 +17,68 @@ class UserCsrData {
+  const certNames = new Set<string>()
+  const pendingNames = new Set<string>()
 
-  try {
-    const start = new Date()
-    response = await fetch(`${serviceAddress}/register`, options)
-    const end = new Date()
-    const fetchTime = (end.getTime() - start.getTime()) / 1000
-    logger(`Fetched ${serviceAddress}, time: ${fetchTime}`)
-  } catch (e) {
-    logger.error(e)
-    return {
-      eventType: RegistrationEvents.ERROR,
-      data: {
-        type: SocketActionTypes.REGISTRAR,
-        code: ErrorCodes.NOT_FOUND,
-        message: ErrorMessages.REGISTRAR_NOT_FOUND,
-        community: communityId,
-      },
+  payload.certificates.forEach(cert => {
+    const parsedCert = parseCertificate(cert)
+    const username = getCertFieldValue(parsedCert, CertFieldsTypes.nickName)
+    if (username) {
+      certNames.add(username)
     }
-  } finally {
-    clearTimeout(timeout)
-  }
+  })
 
-  switch (response?.status) {
-    case 200:
-      break
-    case 400:
-      return {
-        eventType: RegistrationEvents.ERROR,
-        data: {
-          type: SocketActionTypes.REGISTRAR,
-          code: ErrorCodes.BAD_REQUEST,
-          message: ErrorMessages.INVALID_USERNAME,
-          community: communityId,
-        },
-      }
-    case 403:
-      return {
-        eventType: RegistrationEvents.ERROR,
-        data: {
-          type: SocketActionTypes.REGISTRAR,
-          code: ErrorCodes.FORBIDDEN,
-          message: ErrorMessages.USERNAME_TAKEN,
-          community: communityId,
-        },
-      }
-    case 404:
-      return {
-        eventType: RegistrationEvents.ERROR,
-        data: {
-          type: SocketActionTypes.REGISTRAR,
-          code: ErrorCodes.NOT_FOUND,
-          message: ErrorMessages.REGISTRAR_NOT_FOUND,
-          community: communityId,
-        },
-      }
-    default:
-      logger.error(`Registrar responded with ${response?.status} "${response?.statusText}" (${communityId})`)
-      return {
-        eventType: RegistrationEvents.ERROR,
-        data: {
-          type: SocketActionTypes.REGISTRAR,
-          code: ErrorCodes.SERVER_ERROR,
-          message: ErrorMessages.REGISTRATION_FAILED,
-          community: communityId,
-        },
-      }
-  }
+  const pendingCsrs: string[] = []
 
-  const registrarResponse: UserCertificatePayload = await response.json()
+  for (const csr of payload.csrs) {
+    const parsedCsr = await loadCSR(csr)
+    const username = getReqFieldValue(parsedCsr, CertFieldsTypes.nickName)
 
-  logger(`Sending user certificate (${communityId})`)
-  return {
-    eventType: SocketActionTypes.SEND_USER_CERTIFICATE,
-    data: {
-      communityId: communityId,
-      payload: registrarResponse,
-    },
+    if (username && !certNames.has(username) && !pendingNames.has(username)) {
+      pendingNames.add(username)
+      pendingCsrs.push(csr)
+    }
   }
+  return pendingCsrs
```

Comment:
I have one more thought on this. I think we need to take a little more care here. We only want to process a single CSR per public key and we want to process the latest CSR submitted for each public key, yet at the same time, we want to process CSRs in order. Otherwise, if we don't process the latest CSR per public key, then nobody would be able to change names because we would always process their first submitted CSR.

It seems like we need to reverse the CSR list, find the first CSR for each public key (which gives us the most recent CSR for each public key). Then reverse that list so that it is back in order. Then filter CSRs to remove CSRs for usernames that are already taken. Then remove duplicates by taking the first CSR with a specific name.